shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -291,100 +252,173 @@ class RequirementSet {
-    check(other, RequirementSet);
+  return result;
+}
 
-    for (let grainId in other.identityPermissions) {
-      for (let identityId in other.identityPermissions[grainId]) {
-        let otherPermissions = other.identityPermissions[grainId][identityId];
-        this._ensureEntryExists(grainId, identityId);
-        this.identityPermissions[grainId][identityId].add(otherPermissions);
+class Variable {
+  // Our permissions computation can be framed as a propositional HORNSAT problem; this `Variable`
+  // class represents a variable in that sense. There is a variable for every (grain ID, vertex ID,
+  // permission ID) triple. In any given computation, we only explicitly construct those
+  // variables that know might actually be relevant to the result.
+  //
+  // The value of a variable represents an answer to the question "does this vertex in the sharing
+  // graph receive this permission at this grain?" We start out by setting all variables to `false`,
+  // and we only set a variable to `true` when an edge in the sharing graph forces us to. If
+  // this forward-chaining eventually forces us to set our end goal variables to `true`, then the
+  // HORNSAT problem is unsatisfiable and we have proved what we wanted. Otherwise, the HORNSAT
+  // problem is satisfiable, i.e. there is a consistent way to set values to variables in which
+  // our goal nodes do *not* receive the permissions we wanted them to.
+
+  constructor() {
+    this.value = false;
+
+    this.directTailList = [];
+    // List of token IDs for outgoing edges that need to be looked at once this variable gets set
+    // to `true`.
+
+    this.requirementsTailList = [];
+    // List of token IDs for tokens that have requirements that get fulfilled once this variable
+    // gets set to `true`.
+  }
+}
+
+class ActiveToken {
+  // An "active token" is one that we allow to propagate permissions because we've decided
+  // that it might be relevant to our current computation. This class tracks which permissions
+  // the token carries, which of those permissions we've actually proved to arrive at the source
+  // end of the token, and how many of the token's requirements are still unmet.
+
+  constructor(numUnmetRequirements, permissions) {
+    check(numUnmetRequirements, Match.Integer);
+    check(permissions, [Boolean]);
+
+    this.numUnmetRequirements = numUnmetRequirements;
+
+    // The following fields have a "permission status" value, one of:
+    // {unmet: null}, {met: null}, or { doesNotCarry: null }.
+    this.canAccess = { unmet: null };
+    this.appDefined = [];
+
+    for (let ii = 0; ii < permissions.length; ++ii) {
+      if (permissions[ii]) {
+        this.appDefined.push({ unmet: null });
+      } else {
+        this.appDefined.push({ doesNotCarry: null });
       }
     }
   }
-}
 
-class ConditionalPermissionSet {
-  // A PermissionSet that is contingent upon some requirements.
+  requirementsAreMet() {
+    return this.numUnmetRequirements <= 0;
```

Comment:
Presumably a negative value here suggests a horrible bug where some requirements may be being ignored? Maybe `decrementRequirements()` should throw if the count goes negative?